### DISCUSSION TITLE
Pyodbc for latest Python version(irrespective of version)

### DISCUSSION THREAD

**Discussion by marmikreal** [mkleehammer/pyodbc#1113] [Ideas]
PyODBC is designed in such a way that it always does not work with latest python version.

Can we fix that.?

It was working fine with Python 3.10 for me.
But  as I upgraded Python 3.11 it is not working 
Even with backward compatibility it should work fine!

**Comment by marmikreal** [comment]
PS D:\Py\Flask> pip install pyodbc --use-pep517
Collecting pyodbc
  Using cached pyodbc-4.0.34.tar.gz (271 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Building wheels for collected packages: pyodbc
  Building wheel for pyodbc (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Building wheel for pyodbc (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [12 lines of output]
      running bdist_wheel
      running build
      running build_ext
      building 'pyodbc' extension
      creating build
      creating build\temp.win-amd64-cpython-311
      creating build\temp.win-amd64-cpython-311\Release\src
      "C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.33.31629\bin\HostX86\x64\cl.exe" /c /nologo /O2 /W3 /GL /DNDEBUG /MD -DPYODBC_VERSION=4.0.34 -IC:\Users\<user>\AppData\L-IC:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.33.31629\ATLMFC\include" "-IC:\Program Files\Microsoft Visual Studio\2022\Community\VC\Auxiliary\VS\include" "-IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\include\um" /EHsc /Tpsrc\buffer.cpp /Fobuild\temp.win-amd64-cpython-311\Release\src\buffer.obj /Wall /wd4514 /wd4820 /wd4668 /wd4711 /wd4100 /wd4127 /wd4191 /d2FH4-
ory
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for pyodbc
Failed to build pyodbc
ERROR: Could not build wheels for pyodbc, which is required to install pyproject.toml-based projects

**Reply by marmikreal** [reply]
I have tried option --use-pep517 to support backward technology that also fails. with latest 3.11 python version.

**Comment by premins** [comment]
Any info on when the 3.11 version will be available on PIP?

**Reply by gordthompson** [reply]
> Any info on when the 3.11 version will be available on PIP?

Wheels for Python 3.11 are available on PyPI now. (pyodbc 4.0.35, released 2022-11-15)

**Comment by mkleehammer** [comment]
Also, we are looking at using the Python stable API which would allow pyodbc to be used for multiple versions.  It is certainly possible, but I want to look at performance since many macros are not available in the ABI builds.

**Comment by holly-t-123** [comment]
Also wondering about this. I have to upgrade our company apps to python 3.12 for security reasons and now none of them will build because pyodbc is not compatible with 3.12. Any idea on timing for the python stable or python 3.12 update?

**Reply by mkleehammer** [reply]
FYI, 5.0.1 is out with support for Python 3.12

**Comment by mkleehammer** [comment]
I would not have expected this due to 3.12 still being a prerelease version.  The 5.0 version will support Python 3.12 and I plan on releasing it next week.  I just published an alpha (5.0.0a2).  Please see its notes.

I plan on testing the stable API soon after.  I need to setup some performance benchmarks first as the ABI can be slower.  If it is acceptable, it would solve the problem though.

**Comment by alisson-ov-alves** [comment]
Is there a standard way to use pyodbc with Linux and install the necessary drivers?

I use the official python image (python:3.11.4) which has the debian version, but I can't get pyodbc to work.


my dockfile.

FROM python:3.11.4

WORKDIR /app

COPY requirements.txt requirements.txt

RUN apt-get update -q && \
    apt-get install -y g++ unixodbc unixodbc-dev && \ 
    apt-get clean && \
    rm -rf /var/lib/apt/lists/*

ENV LD_LIBRARY_PATH /usr/lib/x86_64-linux-gnu

RUN pip install --upgrade pip

RUN useradd -m -s /bin/bash appexportnetezza

USER appexportnetezza

ENV PATH="/home/appexportnetezza/.local/bin:${PATH}"

RUN pip3 install --user -r requirements.txt

COPY . .

ENTRYPOINT python3 run.py

**Reply by v-chojas** [reply]
What's the error?

**Comment by D00mSlayer** [comment]
I have a similar issue with the following poetry setup:
```
[tool.poetry.dependencies]
python = "3.10.9"
pyodbc = "^4.0.34"
```

fails with

```
  Building wheels for collected packages: pyodbc
    Building wheel for pyodbc (pyproject.toml): started
    Building wheel for pyodbc (pyproject.toml): finished with status 'error'
    error: subprocess-exited-with-error
    
    × Building wheel for pyodbc (pyproject.toml) did not run successfully.
    │ exit code: 1
    ╰─> [18 lines of output]
        WARNING: '' not a valid package name; please use only .-separated package names in setup.py
        running bdist_wheel
        running build
        running build_py
        creating build
        creating build/lib.linux-x86_64-cpython-310
        copying src/pyodbc.pyi -> build/lib.linux-x86_64-cpython-310
        running build_ext
        building 'pyodbc' extension
        creating build/temp.linux-x86_64-cpython-310
        creating build/temp.linux-x86_64-cpython-310/src
        gcc -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DPYODBC_VERSION=4.0.39 -I/home/profile1/.pyenv/versions/3.10.9/envs/flaskace/include -I/home/profile1/.pyenv/versions/3.10.9/include/python3.10 -c src/buffer.cpp -o build/temp.linux-x86_64-cpython-310/src/buffer.o -Wno-write-strings
        In file included from src/buffer.cpp:12:
        src/pyodbc.h:56:10: fatal error: sql.h: No such file or directory
           56 | #include <sql.h>
              |          ^~~~~~~
        compilation terminated.
        error: command '/usr/bin/gcc' failed with exit code 1
        [end of output]
    
    note: This error originates from a subprocess, and is likely not a problem with pip.
    ERROR: Failed building wheel for pyodbc
  Failed to build pyodbc
  ERROR: Could not build wheels for pyodbc, which is required to install pyproject.toml-based projects
```

**Comment by v-chojas** [comment]
Make sure you install unixodbc.
https://github.com/mkleehammer/pyodbc/issues?q=sql.h

**Comment by pbansode** [comment]
working with docker python:3.11-slim-bullseye 

start pyodbc ====================================================================

RUN apt-get update && apt-get -y install apt-transport-https curl
RUN apt-get update && apt-get install -y gnupg2
RUN apt-get update && apt-get install -y software-properties-common --fix-missing
RUN apt-get update -y

RUN apt-get update && apt-get install sudo
RUN apt-get install -y libssl-dev 

RUN sudo apt-get install wget && sudo apt-get update
RUN sudo wget http://archive.ubuntu.com/ubuntu/pool/main/g/glibc/multiarch-support_2.27-3ubuntu1.5_amd64.deb
RUN sudo apt-get install ./multiarch-support_2.27-3ubuntu1.5_amd64.deb


RUN sudo su
RUN curl https://packages.microsoft.com/keys/microsoft.asc | apt-key add - \
    && curl https://packages.microsoft.com/config/debian/9/prod.list > /etc/apt/sources.list.d/mssql-release.list
RUN exit
RUN sudo apt-get update
RUN sudo ACCEPT_EULA=Y apt-get install -y msodbcsql17
RUN sudo ACCEPT_EULA=Y apt-get install -y mssql-tools 
RUN echo 'export PATH="$PATH:/opt/mssql-tools/bin"' >> ~/.bash_profile
RUN echo 'export PATH="$PATH:/opt/mssql-tools/bin"' >> ~/.bashrc
RUN . ~/.bashrc
RUN sudo apt-get update
RUN sudo apt-get -y install unixodbc-dev


RUN sudo apt-get update
RUN sudo apt-get install -y g++
RUN sudo apt-get upgrade -y libstdc++6

RUN sudo apt-get install -y libssl-dev
RUN sudo apt-get install libffi-dev
RUN sudo apt-get update

RUN pip install pyodbc
RUN pip install cython


end pyodbc ====================================================================

pyodbc==5.2.0 under requiremensts.txt
